src/Bicep.Core/Semantics/MetadataSymbol.cs

```diff
@@ -0,0 +1,44 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using System.Collections.Generic;
+using Bicep.Core.Diagnostics;
+using Bicep.Core.Syntax;
+
+namespace Bicep.Core.Semantics
+{
+    public class MetadataSymbol : DeclaredSymbol
+    {
+        public MetadataSymbol(ISymbolContext context, string name, MetadataDeclarationSyntax declaringSyntax, SyntaxBase value)
+            : base(context, name, declaringSyntax, declaringSyntax.Name)
+        {
+            this.Value = value;
+        }
+
+        public SyntaxBase Value { get; }
+
+        public override void Accept(SymbolVisitor visitor) => visitor.VisitMetadataSymbol(this);
+
+        public override SymbolKind Kind => SymbolKind.Metadata;
+
+        public override IEnumerable<Symbol> Descendants
+        {
+            get
+            {
+                yield return this.Type;
+            }
+        }
+
+        public override IEnumerable<ErrorDiagnostic> GetDiagnostics() => ValidateName();
+
+        private IEnumerable<ErrorDiagnostic> ValidateName()
+        {
+            IList<ErrorDiagnostic> diagnostics = new List<ErrorDiagnostic>();
+            if (this.Name.StartsWith("_"))
+            {
+                diagnostics.Add(DiagnosticBuilder.ForPosition(this.NameSyntax).ReservedIdentifier(this.Name));
```

Comment:
I think it is worth creating a new diagnostic for this scenario. `Identifier "_blah" is a reserved Bicep symbol name ...` might cause confusion, as it indicates there's an issue with the specific name chosen rather than a general rule blocking identifiers from starting with `_`.